src/analysis/modifications/moment.jl

```diff
@@ -0,0 +1,110 @@
+"""
+add_moment_constraints(
+    kcats::Dict{String,Float64},
+    molar_weights::Dict{String,Float64},
+    protein_mass_fraction::Float64;
+    protected_constraint_ids = ["ATPM"],
+)
+
+Add enzyme capacity constraints to problem using the MOMENT algorithm. Requires `kcats` and
+`molar_weights` for each reaction and protein. Proteins are encoded by their associated gene.
+Adds a variable vector `y` to the problem corresponding to the protein concentration. Adds a
+variable vector `t` that is internal to the problem - ignore. Set the `protein_mass_fraction`
+capacity constraint to the fraction of the cell mass that is composed of proteins.
+Optionally `protected_constraint_ids` can be used to ensure that some reaction fluxes are not
+unconstrained; all reaction `ids` not protected are set as unconstrained in the direction they
+are set to in the original model. Exchange reactions are protected by default.
+
+See Adadi, Roi, et al. "Prediction of microbial growth rate versus biomass yield by a
+metabolic network with kinetic parameters." PLoS computational biology (2012) for more
+details.
+
+# Example
+```
+sol = flux_balance_analysis_dict(
+    model,
+    Tulip.Optimizer;
+    modifications = [
+        add_moment_constraints(
+            kcats,
+            molar_weights,
+            protein_mass_fraction;
+            protected_constraint_ids = protected_constraint_ids,
+        ),
+        change_constraint("EX_glc__D_e", lb = -1000),
+    ],
+)
+```
+"""
+add_moment_constraints(
+    kcats::Dict{String,Float64},
+    molar_weights::Dict{String,Float64},
+    protein_mass_fraction::Float64;
+    protected_constraint_ids = ["ATPM"],
+) =
+    (model, opt_model) -> begin
+
+        mws = [molar_weights[k] for k in genes(model)] # molar weights in order
+        lbs, ubs = bounds(model) # to assign directions
+        grrs = [reaction_gene_association(model, rid) for rid in reactions(model)]
+
+        # add protein variables
+        y = @variable(opt_model, y[1:n_genes(model)] >= 0)
+
+        # add temporary variables
+        num_temp = sum(length(gids) for grr in grrs for gids in grr) # number "AND" blocks in grrs
+        t = @variable(opt_model, t[1:num_temp] >= 0)
```

Comment:
not all of these will be used because of the "protected" filter. It would probably be much better to add these dynamically in the loop.